Set up Configuration

In [1]:
%load_ext sql

In [2]:
# ID와 PW를 자신의 환경에 맞게 수정(ID:PW 변수 하드코딩 해야함.)
%sql postgresql://ID:PW@HOST:5439/dev

/usr/local/lib/python3.6/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


'Connected: sdrlurker@dev'

Example of Running a SQL:  
이후 아래 SQL에서 스키마는 자신의 스키마로 변경

In [3]:
%%sql

SELECT * FROM raw_data.session_timestamp LIMIT 10

 * postgresql://sdrlurker:***@HOST:5439/dev
10 rows affected.


[('7cdace91c487558e27ce54df7cdb299c', datetime.datetime(2019, 5, 1, 0, 13, 11, 783000)),
 ('94f192dee566b018e0acf31e1f99a2d9', datetime.datetime(2019, 5, 1, 0, 49, 46, 73000)),
 ('7ed2d3454c5eea71148b11d0c25104ff', datetime.datetime(2019, 5, 1, 10, 18, 43, 210000)),
 ('f1daf122cde863010844459363cd31db', datetime.datetime(2019, 5, 1, 13, 10, 56, 413000)),
 ('fd0efcca272f704a760c3b61dcc70fd0', datetime.datetime(2019, 5, 1, 13, 45, 19, 793000)),
 ('8804f94e16ba5b680e239a554a08f7d2', datetime.datetime(2019, 5, 1, 14, 23, 7, 660000)),
 ('c5f441cd5f43eb2f2c024e1f8b5d00cd', datetime.datetime(2019, 5, 1, 15, 3, 54, 650000)),
 ('d5fcc35c94879a4afad61cacca56192c', datetime.datetime(2019, 5, 1, 15, 13, 16, 140000)),
 ('3d191ef6e236bd1b9bdb9ff4743c47fe', datetime.datetime(2019, 5, 1, 15, 33, 58, 197000)),
 ('c17028c9b6e0c5deaad29665d582284a', datetime.datetime(2019, 5, 1, 15, 59, 57, 490000))]

Example of Converting a SQL result into Pandas Dataframe

In [4]:
result = %sql SELECT * FROM raw_data.channel
df = result.DataFrame()

 * postgresql://sdrlurker:***@HOST:5439/dev
7 rows affected.


In [5]:
df.head()

channelname
0     Organic
1      Google
2       Naver
3    Facebook
4     Youtube

# Assignment - 1

## SQL 연습: Channel 테이블 만들어보기(19)
* SQL 연습: Channel 테이블 만들어보기

In [6]:
%%sql

DROP TABLE IF EXISTS sdrlurker.test_channel; -- (2)
CREATE TABLE sdrlurker.test_channel ( 
   channel varchar(32) primary key
); -- (1)

 * postgresql://sdrlurker:***@HOST:5439/dev
Done.
Done.


[]

In [7]:
%%sql

INSERT INTO sdrlurker.test_channel VALUES ('FACEBOOK'), ('GOOGLE'); -- (1)

 * postgresql://sdrlurker:***@HOST:5439/dev
2 rows affected.


[]

In [8]:
%%sql

SELECT * FROM sdrlurker.test_channel; -- (1)

 * postgresql://sdrlurker:***@HOST:5439/dev
2 rows affected.


[('FACEBOOK',), ('GOOGLE',)]

In [9]:
%%sql

ALTER TABLE sdrlurker.test_channel RENAME channel to channelname; -- (3)
INSERT INTO sdrlurker.test_channel VALUES ('TIKTOK'); -- (3)

 * postgresql://sdrlurker:***@HOST:5439/dev
Done.
1 rows affected.


[]

## SQL 연습: IN @ LIKE/ILIKE(22)

In [10]:
%%sql

SELECT COUNT(1) 
FROM raw_data.user_session_channel
WHERE channel in ('Google', 'Facebook');

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[(33773,)]

In [11]:
%%sql

SELECT DISTINCT channel
FROM raw_data.user_session_channel
WHERE channel ILIKE '%o%';

 * postgresql://sdrlurker:***@HOST:5439/dev
4 rows affected.


[('Facebook',), ('Youtube',), ('Organic',), ('Google',)]

In [12]:
%%sql

SELECT COUNT(1) 
FROM raw_data.user_session_channel
WHERE channel ILIKE 'Google' OR
channel ILIKE 'Facebook';

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[(33773,)]

In [13]:
%%sql

SELECT DISTINCT channel
FROM raw_data.user_session_channel
WHERE channel NOT ILIKE '%o%';

 * postgresql://sdrlurker:***@HOST:5439/dev
2 rows affected.


[('Naver',), ('Instagram',)]

## SQL 연습: session이 가장 많이 생성되는 시간대? (27)

In [14]:
%%sql

SELECT EXTRACT(HOUR FROM ts) AS HOUR, COUNT(1) 
FROM raw_data.session_timestamp
GROUP BY 1 
ORDER BY hour;

 * postgresql://sdrlurker:***@HOST:5439/dev
24 rows affected.


[(0, 1870),
 (1, 1328),
 (2, 1125),
 (3, 815),
 (4, 703),
 (5, 611),
 (6, 491),
 (7, 236),
 (8, 149),
 (9, 340),
 (10, 870),
 (11, 1411),
 (12, 3852),
 (13, 9035),
 (14, 13417),
 (15, 15339),
 (16, 14140),
 (17, 10698),
 (18, 7469),
 (19, 5278),
 (20, 3956),
 (21, 3238),
 (22, 2801),
 (23, 2348)]

* 세션이 가장 많이 생성되는 요일
  - 0은 일요일, ... 6이 토요일

In [15]:
%%sql

SELECT EXTRACT(DOW FROM ts) AS dow, COUNT(1) 
FROM raw_data.session_timestamp
GROUP BY 1 
ORDER BY dow;

 * postgresql://sdrlurker:***@HOST:5439/dev
7 rows affected.


[(0, 13074),
 (1, 14625),
 (2, 14762),
 (3, 14999),
 (4, 14901),
 (5, 15091),
 (6, 14068)]

## SQL 연습: raw_data.channel의 채널별 사용자수 세기 (32)

In [16]:
%%sql

SELECT * FROM raw_data.channel LIMIT 10;

 * postgresql://sdrlurker:***@HOST:5439/dev
7 rows affected.


[('Naver',),
 ('Facebook',),
 ('Youtube',),
 ('Instagram',),
 ('TIKTOK',),
 ('Organic',),
 ('Google',)]

In [17]:
%%sql

SELECT * FROM raw_data.user_session_channel LIMIT 10;

 * postgresql://sdrlurker:***@HOST:5439/dev
10 rows affected.


[(184, 'c41dd99a69df04044aa4e33ece9c9249', 'Naver'),
 (80, 'fdc0eb412a84fa549afe68373d9087e9', 'Organic'),
 (251, '0a54b19a13b6712dc04d1b49215423d8', 'Facebook'),
 (264, 'a914ecef9c12ffdb9bede64bb703d877', 'Google'),
 (744, '05ae14d7ae387b93370d142d82220f1b', 'Facebook'),
 (229, 'eddeb82df22554fa67c641e3f8a25566', 'Google'),
 (265, '4c4ea5258ef3fb3fb1fc48fee9b4408c', 'Naver'),
 (80, '8ae1da0fe37c98412768453f82490da2', 'Organic'),
 (45, '60131a2a3f223dc8f4753bcc5771660c', 'Youtube'),
 (198, '90f4760fcc9b69c13da7368c5c2917f3', 'Facebook')]

## SQL 연습: raw_data.channel의 채널별 사용자수 세기 (32) -- answer!

In [18]:
%%sql

SELECT channelname, COUNT(DISTINCT userid)
FROM raw_data.channel C LEFT JOIN raw_data.user_session_channel USC 
ON C.channelname = USC.channel
GROUP BY channelname;

 * postgresql://sdrlurker:***@HOST:5439/dev
7 rows affected.


[('Organic', 895),
 ('Google', 893),
 ('Instagram', 895),
 ('Naver', 882),
 ('Youtube', 889),
 ('Facebook', 889),
 ('TIKTOK', 0)]

## SQL 연습: 사용자별로 처음 채널과 마지막 채널 알아내기 (41)
* 사용자 251번의 시간순으로 봤을 때 첫 번째 채널과 마지막 채널은?
  - 노가다로 아래 쿼리를 실행해서 첫 번째 채널 및 마지막 채널 확인

In [19]:
result = %sql SELECT ts, channel FROM raw_data.user_session_channel USC JOIN raw_data.session_timestamp ST ON USC.sessionid = ST.sessionid WHERE userid=251 ORDER BY 1;
df = result.DataFrame()

 * postgresql://sdrlurker:***@HOST:5439/dev
216 rows affected.


In [20]:
df.head()

ts    channel
0 2019-05-01 12:18:26.637   Facebook
1 2019-05-02 13:06:12.047  Instagram
2 2019-05-03 13:43:21.857  Instagram
3 2019-05-04 13:54:00.610  Instagram
4 2019-05-05 13:30:08.260     Google

In [21]:
df.tail()

ts   channel
211 2019-11-21 15:44:01     Naver
212 2019-11-22 17:55:35  Facebook
213 2019-11-23 15:34:30    Google
214 2019-11-25 14:31:24   Organic
215 2019-11-26 16:54:19    Google

## SQL 연습: 사용자별로 처음 채널과 마지막 채널 알아내기 (41) -- answer!


* ROW_NUMBER 이용해서 해보자
  - 사용자 251번의 첫번째와 마지막 채널 확인.

In [22]:
%%sql

SELECT userid, channel, ROW_NUMBER() OVER (PARTITION BY userid ORDER BY ts) as row_num, ts
FROM raw_data.user_session_channel USC 
JOIN raw_data.session_timestamp ST ON USC.sessionid = ST.sessionid
WHERE userid = 251
ORDER BY row_num ASC
LIMIT 1;

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[(251, 'Facebook', 1, datetime.datetime(2019, 5, 1, 12, 18, 26, 637000))]

In [23]:
%%sql

SELECT userid, channel, ROW_NUMBER() OVER (PARTITION BY userid ORDER BY ts) as row_num, ts
FROM raw_data.user_session_channel USC 
JOIN raw_data.session_timestamp ST ON USC.sessionid = ST.sessionid
WHERE userid = 251
ORDER BY row_num DESC
LIMIT 1;

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[(251, 'Google', 216, datetime.datetime(2019, 11, 26, 16, 54, 19))]

* CTAS로 결과를 user_channel_summary 테이블로 만들었습니다.

In [24]:
%%sql

DROP TABLE IF EXISTS sdrlurker.user_channel_summary;

CREATE TABLE sdrlurker.user_channel_summary AS 
SELECT userid, first, last, COUNT(1), min(ts), max(ts) FROM 
(SELECT DISTINCT userid, 
FIRST_VALUE(channel) OVER (PARTITION BY userid ORDER BY ts ROWS UNBOUNDED PRECEDING) AS first, 
FIRST_VALUE(channel) OVER (PARTITION BY userid ORDER BY ts DESC ROWS UNBOUNDED PRECEDING) AS last,
ROW_NUMBER() OVER (PARTITION BY userid ORDER BY ts) as row_num,
ts
FROM raw_data.user_session_channel USC 
JOIN raw_data.session_timestamp ST ON USC.sessionid = ST.sessionid
ORDER BY userid) tmp 
GROUP BY 1, 2, 3
ORDER BY userid;

 * postgresql://sdrlurker:***@HOST:5439/dev
Done.
Done.


[]

* userid 251의 처음 채널은 Facebook, 마지막 채널은 Google, 개수 216, 시간대가 맞음을 검증합니다.

In [25]:
%%sql

SELECT *
FROM sdrlurker.user_channel_summary
WHERE userid = 251;

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[(251, 'Facebook', 'Google', 216, datetime.datetime(2019, 5, 1, 12, 18, 26, 637000), datetime.datetime(2019, 11, 26, 16, 54, 19))]

# Assignment - 2


## Gross Revenue가 가장 큰 UserID 10개 찾기

In [26]:
%%sql

SELECT userid, SUM(amount) as gross_revenue
FROM raw_data.user_session_channel USC 
LEFT JOIN raw_data.session_transaction ST ON USC.sessionid = ST.sessionid
GROUP BY 1
ORDER BY gross_revenue DESC NULLS LAST
LIMIT 10;

 * postgresql://sdrlurker:***@HOST:5439/dev
10 rows affected.


[(989, 743),
 (772, 556),
 (1615, 506),
 (654, 488),
 (1651, 463),
 (973, 438),
 (262, 422),
 (1099, 421),
 (2682, 414),
 (891, 412)]

# Assignment - 3


## 채널별 월 매출액 테이블 만들기 
* conversionRate 소수점 표기 방법
  - https://stackoverflow.com/questions/50494508/redshift-division-result-does-not-include-decimals

In [27]:
%%sql

SELECT channel, DATE_TRUNC('month', ts) AS month,
COUNT(DISTINCT userid) AS uniqueUsers, 
COUNT(DISTINCT (CASE WHEN refunded IS NOT NULL then userid END)) AS paidUsers,
CASE WHEN uniqueUsers=0 then NULL ELSE 1.0 * paidUsers / uniqueUsers END conversionRate,
SUM(CASE WHEN refunded IS NOT NULL then amount END) AS grossRevenue,
SUM(CASE WHEN refunded IS FALSE then amount END) AS netRevenue
FROM raw_data.channel C 
LEFT JOIN raw_data.user_session_channel USC
ON C.channelname = USC.channel
LEFT JOIN raw_data.session_timestamp ST 
ON USC.sessionid = ST.sessionid
LEFT JOIN raw_data.session_transaction ST2
ON USC.sessionid = ST2.sessionid
GROUP BY 1, 2
ORDER BY channel, month;

 * postgresql://sdrlurker:***@HOST:5439/dev
43 rows affected.


[('Facebook', datetime.datetime(2019, 5, 1, 0, 0), 247, 14, Decimal('0.05668016194331983'), 1199, 997),
 ('Facebook', datetime.datetime(2019, 6, 1, 0, 0), 414, 22, Decimal('0.05314009661835748'), 1578, 1578),
 ('Facebook', datetime.datetime(2019, 7, 1, 0, 0), 558, 32, Decimal('0.05734767025089605'), 2222, 2144),
 ('Facebook', datetime.datetime(2019, 8, 1, 0, 0), 611, 18, Decimal('0.02945990180032733'), 1009, 1009),
 ('Facebook', datetime.datetime(2019, 9, 1, 0, 0), 597, 27, Decimal('0.04522613065326633'), 2270, 2270),
 ('Facebook', datetime.datetime(2019, 10, 1, 0, 0), 698, 29, Decimal('0.04154727793696275'), 1650, 1641),
 ('Facebook', datetime.datetime(2019, 11, 1, 0, 0), 688, 26, Decimal('0.03779069767441860'), 1678, 1678),
 ('Google', datetime.datetime(2019, 5, 1, 0, 0), 253, 10, Decimal('0.03952569169960474'), 580, 580),
 ('Google', datetime.datetime(2019, 6, 1, 0, 0), 412, 13, Decimal('0.03155339805825242'), 947, 947),
 ('Google', datetime.datetime(2019, 7, 1, 0, 0), 556, 21, Decimal('0.03776978417266187'), 1558, 1385),
 ('Google', datetime.datetime(2019, 8, 1, 0, 0), 610, 28, Decimal('0.04590163934426229'), 2210, 1894),
 ('Google', datetime.datetime(2019, 9, 1, 0, 0), 599, 25, Decimal('0.04173622704507512'), 1872, 1691),
 ('Google', datetime.datetime(2019, 10, 1, 0, 0), 699, 30, Decimal('0.04291845493562231'), 2150, 2098),
 ('Google', datetime.datetime(2019, 11, 1, 0, 0), 688, 26, Decimal('0.03779069767441860'), 2286, 2235),
 ('Instagram', datetime.datetime(2019, 5, 1, 0, 0), 234, 11, Decimal('0.04700854700854700'), 959, 770),
 ('Instagram', datetime.datetime(2019, 6, 1, 0, 0), 410, 21, Decimal('0.05121951219512195'), 1462, 1418),
 ('Instagram', datetime.datetime(2019, 7, 1, 0, 0), 567, 25, Decimal('0.04409171075837742'), 1896, 1766),
 ('Instagram', datetime.datetime(2019, 8, 1, 0, 0), 621, 28, Decimal('0.04508856682769726'), 2129, 2001),
 ('Instagram', datetime.datetime(2019, 9, 1, 0, 0), 588, 20, Decimal('0.03401360544217687'), 1260, 1122),
 ('Instagram', datetime.datetime(2019, 10, 1, 0, 0), 707, 34, Decimal('0.04809052333804809'), 2568, 2395),
 ('Instagram', datetime.datetime(2019, 11, 1, 0, 0), 669, 25, Decimal('0.03736920777279521'), 2116, 2116),
 ('Naver', datetime.datetime(2019, 5, 1, 0, 0), 237, 11, Decimal('0.04641350210970464'), 867, 844),
 ('Naver', datetime.datetime(2019, 6, 1, 0, 0), 398, 15, Decimal('0.03768844221105527'), 1090, 1090),
 ('Naver', datetime.datetime(2019, 7, 1, 0, 0), 553, 19, Decimal('0.03435804701627486'), 1547, 1547),
 ('Naver', datetime.datetime(2019, 8, 1, 0, 0), 626, 22, Decimal('0.03514376996805111'), 1829, 1551),
 ('Naver', datetime.datetime(2019, 9, 1, 0, 0), 592, 21, Decimal('0.03547297297297297'), 1996, 1996),
 ('Naver', datetime.datetime(2019, 10, 1, 0, 0), 713, 32, Decimal('0.04488078541374474'), 2695, 2695),
 ('Naver', datetime.datetime(2019, 11, 1, 0, 0), 667, 26, Decimal('0.03898050974512743'), 2234, 1987),
 ('Organic', datetime.datetime(2019, 5, 1, 0, 0), 238, 17, Decimal('0.07142857142857142'), 1846, 1571),
 ('Organic', datetime.datetime(2019, 6, 1, 0, 0), 416, 14, Decimal('0.03365384615384615'), 1129, 940),
 ('Organic', datetime.datetime(2019, 7, 1, 0, 0), 557, 24, Decimal('0.04308797127468581'), 1600, 1600),
 ('Organic', datetime.datetime(2019, 8, 1, 0, 0), 608, 26, Decimal('0.04276315789473684'), 1643, 1606),
 ('Organic', datetime.datetime(2019, 9, 1, 0, 0), 592, 23, Decimal('0.03885135135135135'), 1267, 1267),
 ('Organic', datetime.datetime(2019, 10, 1, 0, 0), 709, 31, Decimal('0.04372355430183356'), 2762, 2608),
 ('Organic', datetime.datetime(2019, 11, 1, 0, 0), 677, 34, Decimal('0.05022156573116691'), 2626, 2255),
 ('Youtube', datetime.datetime(2019, 5, 1, 0, 0), 244, 10, Decimal('0.04098360655737704'), 529, 529),
 ('Youtube', datetime.datetime(2019, 6, 1, 0, 0), 400, 17, Decimal('0.04250000000000000'), 1042, 1042),
 ('Youtube', datetime.datetime(2019, 7, 1, 0, 0), 564, 36, Decimal('0.06382978723404255'), 2210, 2037),
 ('Youtube', datetime.datetime(2019, 8, 1, 0, 0), 

In [28]:
%%sql

DROP TABLE IF EXISTS sdrlurker.monthly_channel_revenue;
CREATE TABLE sdrlurker.monthly_channel_revenue AS 
SELECT channel, DATE_TRUNC('month', ts) AS month,
COUNT(DISTINCT userid) AS uniqueUsers, 
COUNT(DISTINCT (CASE WHEN refunded IS NOT NULL then userid END)) AS paidUsers,
CASE WHEN uniqueUsers=0 then NULL ELSE 1.0 * paidUsers / uniqueUsers END conversionRate,
SUM(CASE WHEN refunded IS NOT NULL then amount END) AS grossRevenue,
SUM(CASE WHEN refunded IS FALSE then amount END) AS netRevenue
FROM raw_data.channel C 
LEFT JOIN raw_data.user_session_channel USC
ON C.channelname = USC.channel
LEFT JOIN raw_data.session_timestamp ST 
ON USC.sessionid = ST.sessionid
LEFT JOIN raw_data.session_transaction ST2
ON USC.sessionid = ST2.sessionid
GROUP BY 1, 2
ORDER BY channel, month;

 * postgresql://sdrlurker:***@HOST:5439/dev
Done.
Done.


[]

* Facebook Channel 5월 총 방문자 수 검증합니다. (247)

In [29]:
%%sql

SELECT channel, DATE_TRUNC('month', ts) AS month, count(DISTINCT userid)
FROM raw_data.channel C 
LEFT JOIN raw_data.user_session_channel USC
ON C.channelname = USC.channel
LEFT JOIN raw_data.session_timestamp ST 
ON USC.sessionid = ST.sessionid
LEFT JOIN raw_data.session_transaction ST2
ON USC.sessionid = ST2.sessionid
WHERE channel = 'Facebook' AND ts BETWEEN '2019-05-01 00:00:00' AND '2019-05-31 23:59:59' 
GROUP BY 1, 2;

 * postgresql://sdrlurker:***@HOST:5439/dev
1 rows affected.


[('Facebook', datetime.datetime(2019, 5, 1, 0, 0), 247)]

* Facebook Channel 5월 구매 사용자 수 검증합니다. (14)

In [30]:
%%sql

SELECT channel, ts, userid, refunded, amount
FROM raw_data.channel C 
LEFT JOIN raw_data.user_session_channel USC
ON C.channelname = USC.channel
LEFT JOIN raw_data.session_timestamp ST 
ON USC.sessionid = ST.sessionid
LEFT JOIN raw_data.session_transaction ST2
ON USC.sessionid = ST2.sessionid
WHERE channel = 'Facebook' AND ts BETWEEN '2019-05-01 00:00:00' AND '2019-05-31 23:59:59' 
AND refunded is NOT NULL 
ORDER BY refunded, amount;

 * postgresql://sdrlurker:***@HOST:5439/dev
14 rows affected.


[('Facebook', datetime.datetime(2019, 5, 2, 17, 47, 43), 267, False, 16),
 ('Facebook', datetime.datetime(2019, 5, 15, 18, 28, 11, 910000), 260, False, 29),
 ('Facebook', datetime.datetime(2019, 5, 22, 13, 45, 43, 537000), 676, False, 43),
 ('Facebook', datetime.datetime(2019, 5, 10, 16, 34, 9, 313000), 768, False, 63),
 ('Facebook', datetime.datetime(2019, 5, 18, 17, 13, 10, 393000), 87, False, 105),
 ('Facebook', datetime.datetime(2019, 5, 15, 22, 26, 42), 227, False, 107),
 ('Facebook', datetime.datetime(2019, 5, 2, 15, 15, 47, 890000), 468, False, 111),
 ('Facebook', datetime.datetime(2019, 5, 23, 16, 33, 27, 20000), 772, False, 127),
 ('Facebook', datetime.datetime(2019, 5, 30, 14, 42, 13, 383000), 1139, False, 129),
 ('Facebook', datetime.datetime(2019, 5, 4, 19, 45, 36, 147000), 533, False, 129),
 ('Facebook', datetime.datetime(2019, 5, 20, 21, 19, 20, 753000), 983, False, 138),
 ('Facebook', datetime.datetime(2019, 5, 25, 18, 24, 33), 945, True, 3),
 ('Facebook', datetime.datetime(2019, 5, 15, 20, 46, 54, 207000), 830, True, 80),
 ('Facebook', datetime.datetime(2019, 5, 20, 13, 33, 49, 767000), 225, True, 119)]

* Facebook Channel 5월 구매 grossRevenue를 검증합니다. (1199)

In [31]:
16+29+43+63+105+107+111+127+129+129+138+3+80+119

1199

* Facebook Channel 5월 구매 netRevenue를 검증합니다. (997)

In [32]:
16+29+43+63+105+107+111+127+129+129+138

997